### ISSUE TITLE
Expression documentation feedback

### ISSUE THREAD

**Issue by stevage** [mapbox/mapbox-gl-js#5500]
Awesome to see the new expression functionality released :) I wanted to quickly give some feedback on the documentation before I become too familiar with it, and lose the ability to see what's hard for newcomers to grasp. Apologies for the annoying tone.

- The word "expression" is used a little bit loosely earlier in the doc ("Filter: A expression specifying conditions on source features")
- There are no signposts to the new Expression syntax. It just appears as its own section, and there's no mention of the relationship between Expressions and Functions. A sentence like "In this way, they perform a similar role to Functions, but with more flexibility" would help...if that's true. :)
- I  found the first property expression example a little bit too terse, and took quite a lot of staring at it to understand. I *think* the 0 is referring to the "g" component of the rgb color, but this isn't made clear at all. It seems wonderfully magical that you can directly address components of the colour space this way. There is a lot being introduced in this one example: expression syntax, color space addressing, "get" functions, "-" operator...maybe too much.
- It's surprising, and a bit mysterious, to me that "zoom" is written in square brackets. The reason for that isn't spelled out. (I assume there are reasons why `"$zoom"` wasn't chosen...)
- I found the zoom-and-property expression really hard to take in on first glance. We have an array of 7 items, with unusual line breaks...it doesn't feel very intuitive why those particular 7 items, or how I would know how to construct my own array. "curve" is not in [], but "linear" and "zoom" are. Why?
- Under "Types", I found I just had no context for how to use this information "`array` asserts that the input is an array". I've read that paragraph several times and I just don't know what it's saying, or whether it's useful to me.  In my understanding of Mapbox and expressions to this point, I don't know where "asserting that the input value is X" fits in. (Again, speaking with very fresh eyes here - hopefully this comment will sound stupid to me in a couple of days.)
- Lookup > get: I wonder what else one could query if not the "properties" object?
- Decision: I had trouble understanding where "decisions" fit in. There's been no mention of them somewhere. I understand that these are boolean operators, but we haven't seen an "if" function or equivalent so far, so I'm not sure what I would do with a boolean.  [I guess from further reading, it's used with "case"...]
- Decision > !=: I had trouble understanding what the `["!=", null, null]: boolean` line means. Ok, if I pass two nulls to the `!=` operator, I get a boolean back, but what are the semantics? True or false? Similarly, what are the semantics of `<` (and similar) for string comparisons?
- Decision > all: I found the `(boolean, boolean, ...)` notation a bit confusing. I assume it just means 3 or more booleans, but I don't quite know where there are parentheses there.
- Decision > match: I found the definition too hard to parse. I really couldn't work out what to do with the | and the T and the U. If there are several distinct forms, maybe providing them completely separately, rather than combining them into one definition would be easier. An example would help.
- Decision > curve: Again, very hard to parse. I wasn't clear on what "input" would be. The name of a property? 
- String: Sad that there are no substring/split operations :)
- Color: Sad that hsl/hsv is not supported (yet)...

**Comment by jingsam** [comment]
Yeah, the new expression is making mapbox-gl-js more expressive and data-driven. The problem is that the expression syntax is too verbose and hard to grasp. I feel you guys are inventing some programing language like "Mapbox Lisp".... 😂

**Comment by anandthakker** [comment]
@stevage thank you so much for taking the time to dig in early and collect this feedback!!   We're definitely aware that the documentation right now is _extremely_ minimal, and we're actively working both on expanding these reference docs and writing more guided, explanatory tutorials and guides.  These notes from your fresh, still-unfamiliar perspective will be very helpful!

cc @colleenmcginnis

**Comment by stevage** [comment]
Awesome. Btw I should have said, in general I have an extremely high opinion of the Mapbox-GL-JS documentation - I often point it out to people as examples of what documentation can and should be. The structure of examples/API/style-spec works really well for me - pretty much every Mapbox project starts with copy-pasting one of the examples, usually the first one :) 

I don't particularly use tutorials or guides. So for me personally, I would prefer effort on keeping the main API/style doco excellent and readable (and with really sharp examples in the doco), rather than trying to somehow compensate with layers of tutorials.

**Comment by colleenmcginnis** [comment]
I'm transferring this issue to the new home for the Mapbox GL JS documentation ([/mapbox-gl-js-docs](https://github.com/mapbox/mapbox-gl-js-docs)). See https://github.com/mapbox/mapbox-gl-js/pull/8582 for details on this move.
